### PR TITLE
pass gregor notices up to electron; allow for a pull or push-based workflow

### PR DESCRIPTION
**Pull request by maxtaco**

### CODE REVIEW

**Comment by maxtaco** [comment]
cc: @chrisnojima, let's hack on this when you get a chance....

**Comment by chrisnojima** [comment]
My PR fixes the js stuff: https://github.com/keybase/client/pull/3091

**Comment by chrisnojima** [comment]
my pr was merged down

**Comment by maxtaco** [comment]
@mmaxim can you help me to review this?  the guts are rather simple but I've had to pull in a bunch of stuff.

File: go/engine/signup.go [hunk 1]
```diff
@@ -114,6 +114,9 @@ func (s *SignupEngine) Run(ctx *Context) error {
 	// a chance of timing out)
 	s.G().NotifyRouter.HandleLogin(s.arg.Username)
 
+	// For instance, setup gregor and friends...
+	s.G().CallLoginHooks()
```
Comment (maxtaco) [inline go/engine/signup.go]: this was the left out login hook

File: go/service/gregor.go [hunk 2]
```diff
@@ -697,38 +721,78 @@ func (g *gregorHandler) connectNoTLS(uri *rpc.FMPURI) error {
...
+	}
+
 	incomingClient := gregor1.IncomingClient{Cli: g.cli}
 	// TODO: Should the interface take a context from the caller?
-	return incomingClient.ConsumeMessage(context.TODO(), dismissal)
+	err = incomingClient.ConsumeMessage(context.TODO(), *dismissal)
+	return err
+}
+
+func (g *gregorHandler) InjectItem(cat string, body []byte) (gregor.MsgID, error) {
```
Comment (maxtaco) [inline go/service/gregor.go]: useful for testing

File: go/service/gregor.go [hunk 3]
```diff
@@ -323,6 +323,27 @@ func (g *gregorHandler) ShouldRetryOnConnect(err error) bool {
...
+	if ui == nil {
+		g.Debug("no-op on message, since no Gregor UIs are registerd")
+		return
+	}
+	err = ui.PushMessage(context.Background(), []gregor1.Message{m})
+	if err != nil {
+		g.Warning("Error pushing message to gregor UI: %s", err)
+	}
+	return
+}
```
Comment (mmaxim) [inline go/service/gregor.go]: Probably should drop the use of "gregor" in favor of "push" to be consistent with other messages.

File: go/service/gregor.go [hunk 4]
```diff
@@ -332,6 +353,9 @@ func (g *gregorHandler) BroadcastMessage(ctx context.Context, m gregor1.Message)
 	// Send message to local state machine
 	g.gregorCli.StateMachineConsumeMessage(m)
 
+	// Forward to electron or whichever UI is listening for gregor updates
+	g.forwardMessage(m)
```
Comment (mmaxim) [inline go/service/gregor.go]: What about during a sync, do we want to also forward those?

**Comment by keybase-ci-visdiff** [comment]
:mag_right: The commits 54db08097875...e4d657257baf introduced some visual changes:
 * Changed: **unlock_folders-Success** [(before)](http://keybase-app-visdiff.s3.amazonaws.com/1466115050510-54db08097875-e4d657257baf/unlock_folders-Success-54db08097875.png) [(after)](http://keybase-app-visdiff.s3.amazonaws.com/1466115050510-54db08097875-e4d657257baf/unlock_folders-Success-e4d657257baf.png)  
   unlock_folders-Success comparison
 * Changed: **Passphrase-Error** [(before)](http://keybase-app-visdiff.s3.amazonaws.com/1466115050510-54db08097875-e4d657257baf/Passphrase-Error-54db08097875.png) [(after)](http://keybase-app-visdiff.s3.amazonaws.com/1466115050510-54db08097875-e4d657257baf/Passphrase-Error-e4d657257baf.png)  
   Passphrase-Error comparison

**Comment by maxtaco** [comment]
@cjb, I need an assist here.  I need to set:

```
KEYBASE_PUSH_SERVER_URI=fmprc://api0.ci.keybase.io:9911
```

And also I need to expose that port on `api0.ci`

Not urgent....

Also @mmaxim, this might be worth another look.  I changed the mechanism for plugging into the gregor handler, but still aren't replaying gregor messages on sync. We should chat with @chrisnojima tomorrow to see if this makes sense...

**Comment by maxtaco** [comment]
Also, @cjb, I have no idea how these tests passed Windows CI. They should have failed for the same reason they failed OSX.   Is it possible Windows isn't reporting errors properly?

**Comment by keybase-ci-visdiff** [comment]
:mag_right: The commits 51c89290fcd9...a5c3a0caf13f introduced some visual changes:
 * Changed: **Request_Invite-Start** [(before)](http://keybase-app-visdiff.s3.amazonaws.com/1466198231056-51c89290fcd9-a5c3a0caf13f/Request_Invite-Start-51c89290fcd9.png) [(after)](http://keybase-app-visdiff.s3.amazonaws.com/1466198231056-51c89290fcd9-a5c3a0caf13f/Request_Invite-Start-a5c3a0caf13f.png)  
   Request_Invite-Start comparison
 * Changed: **Profile-Changed** [(before)](http://keybase-app-visdiff.s3.amazonaws.com/1466198231056-51c89290fcd9-a5c3a0caf13f/Profile-Changed-51c89290fcd9.png) [(after)](http://keybase-app-visdiff.s3.amazonaws.com/1466198231056-51c89290fcd9-a5c3a0caf13f/Profile-Changed-a5c3a0caf13f.png)  
   Profile-Changed comparison

**Comment by keybase-ci-visdiff** [comment]
:mag_right: The commits 97dbe3edccac...484aa8032a52 introduced some visual changes:
 * Changed: **Device_Name-Start** [(before)](http://keybase-app-visdiff.s3.amazonaws.com/1466201439786-97dbe3edccac-484aa8032a52/Device_Name-Start-97dbe3edccac.png) [(after)](http://keybase-app-visdiff.s3.amazonaws.com/1466201439786-97dbe3edccac-484aa8032a52/Device_Name-Start-484aa8032a52.png)  
   Device_Name-Start comparison

**Pull request merged**